curve25519-dalek/src/backend/serial/u64/field.rs

```diff
@@ -362,6 +362,24 @@ impl FieldElement51 {
         ])
     }
 
+    /// Load a `FieldElement51` from 64 bytes, by reducing modulo q.
+    pub fn from_bytes_wide(hash: &[u8; 64]) -> FieldElement51 {
+        let mut fl = [0u8; 32];
+        let mut gl = [0u8; 32];
+        fl.copy_from_slice(&hash[..32]);
+        gl.copy_from_slice(&hash[32..]);
+        fl[31] &= 0x7f;
+        gl[31] &= 0x7f;
+
+        let mut fe_f = Self::from_bytes(&fl);
+        let fe_g = Self::from_bytes(&gl);
+        fe_f.0[0] = fe_f.0[0] + (hash[31] >> 7) as u64 * 19 + (hash[63] >> 7) as u64 * 722;
```

Comment:
Where does this math come from?

Reply:
I added some comments that explain it all, but I have one last question: why does the `38 * ` expression below guaranteed to not overflow? Is there a property in libsodium that guarantees this? Do we have the same?

Reply:
`FieldElement51` represents an element of the field `2^255 - 19`. In other words, it represents a number of size 255 bits. With 4 limbs of `u64`, we would have enough for representing these numbers. However, `FieldElement51` represents the number with 5 limbs of 51 bits, and as [documented](https://github.com/dalek-cryptography/curve25519-dalek/blob/main/curve25519-dalek/src/backend/serial/u64/field.rs#L30): 
```
the coefficients are allowed to grow up to 2^54 between reductions modulo p
```

So even if the type is u64, we can do these operations here with the guarantee that we won't have an overflow.

Reply:
If `from_bytes` produces an element with a limb x = 2^51, then isn’t it the case that x * 38 > x * 32 = 2^56 > 2^54?

Reply:
Ah it looks like you indeed can pass limbs up to 2^64

https://github.com/dalek-cryptography/curve25519-dalek/blob/dd5bd108d6985491acb3de25497fb082a29c9fb7/curve25519-dalek/src/backend/serial/u64/field.rs#L293

So the current impl should work. Could you add a comment saying this when you get a chance?